### DISCUSSION TITLE
Does the position of @run_after matter during decoration?

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#3833] [Q&A] [answered]
@alalazo 

I was experimenting with adding an optional post-installation test to a package and noticed the following behavior.
```python
@run_after('install')                                                       
@on_package_attributes(run_tests=True)                                      
def check_install(self):
```
works as expected. The test is run post-installation, but only when `--run-tests` is specified. However, if I do:
```python
@on_package_attributes(run_tests=True)      
@run_after('install')                                                                                       
def check_install(self):
```
instead, the tests are always run regardless of whether or not I supply `--run-tests`.

**Comment by alalazo** [comment, marked as answer]
@adamjstewart That's because when you call `@run_after` or `@run_before` you register the function to be called back, for instance:
```python
def run_after(*phases):
    """Registers a method of a package to be run after a given phase"""
    return PackageMeta.register_callback('run_after', *phases)
```

Now, if you "unroll" the second case to its semantically equivalent:
```
def check_install(self):
   pass

check = run_after('install')(check)  # Gets registered here
check = on_package_attribute(run_tests=True)(check)  # ... but also gets decorated after the registration
```
you see that the second decorator has no effect on the registered function. Maybe worth documenting?

**Comment by alalazo** [comment]
@adamjstewart I changed the title to reflect better what is being discussed here. Feel free to change it back if you disagree.

**Comment by adamjstewart** [comment]
I think it's a bit too complicated to document this one. There's no way to just make it work?

**Comment by adamjstewart** [comment]
@alalazo What if we changed the interface to:
```python
@run_after('install', run_tests=True)                                                       
def check_install(self):
```
I'm not sure how the implementation details would work, but that would avoid a lot of unnecessary confusion.
